src/components/__tests__/page-details.test.jsx

```diff
@@ -0,0 +1,62 @@
+/* eslint-env jest */
+import PageDetails from '../page-details';
+import React from 'react';
+import {shallow} from 'enzyme';
+import simplePage from '../../__mocks__/simple-page.json';
+import WebMonitoringDb from '../../services/web-monitoring-db';
+
+describe('page-details', () => {
+  // Change string values to date objects so they're parsed correctly
+  simplePage.versions.forEach(version => {
+    version.capture_time = new Date(version.capture_time);
+  });
+  const match = { params: { pageId: 'samplePageId' }};
+  const mockApi = Object.assign(Object.create(WebMonitoringDb.prototype), {
+    getDiff () {
+      return new Promise(resolve => resolve({
+        change_count: 1,
+        diff: [[0, 'Hi'], [1, '!']]
+      }));
+    },
+    getPage (a) {
+      return new Promise(resolve => resolve(simplePage));
+    },
+    getVersions (a, b, c) {
+      return new Promise(resolve => resolve(simplePage.versions));
+    },
+    annotateChange (a, b, c, d) {
+      return 'something';
+    }
+  });
+
+  it('can render', () => {
+    const pageDetails = shallow(
+      <PageDetails
+        match={match}
+      />,
+      {context: {api: mockApi}}
+    );
+
+    expect(pageDetails.exists()).toEqual(true);
+    // It should always have rendered *something.*
+    expect(pageDetails.get(0)).toBeTruthy();
+  });
+
+  it('shows correct title', () => {
+    const pageDetails = shallow(
+      <PageDetails
+        match={match}
+      />,
+      {context: {api: mockApi}}
+    );
+    const instance = pageDetails.instance();
+    instance.state.page = simplePage;
+    instance.componentDidUpdate(instance.props);
```

Comment:
You should avoid reaching in and manually setting state and calling lifecycle methods here — when you do, you’re not really testing that the component is functioning properly and doing all the things it should in response to having its props set.

It’s more complicated, but a better approach here is probably something like what @steryereo did for the login form here, where we wait for the mock data to load: https://github.com/edgi-govdata-archiving/web-monitoring-ui/blob/24fe977a334ce1bc3f84568cf5ffe75c771aa4ee/src/components/__tests__/login-form-test.jsx#L77-L80

A version of that here might look like:

```js
// Note this test is *async*
it('shows correct title', async () => {
  // The mock needs to be inside this function so it doesn't get shard with the other test
  const mockApi = Object.assign(Object.create(WebMonitoringDb.prototype), {
    // Jest has some tooling for mock async functions we can use to simplify the code:
    getPage: jest.fn().mockResolvedValue(simplePage),
    getVersions: jest.fn().mockResolvedValue(simplePage.versions),
    getDiff: jest.fn().mockResolvedValue({
      change_count: 1,
      diff: [[0, 'Hi'], [1, '!']]
    })
  });

  // Mount the component, same as before:
  const pageDetails = shallow(
    <PageDetails
      match={match}
    />,
    {context: {api: mockApi}}
  );

  // Now wait for the component to have loaded the data!
  // Since we used `jest.fn()` to create a mock function for `getPage()`,
  // we can retrieve the return value from when the component called it.
  // Since the return value is a promise, we just wait for it to finish:
  await mockApi.getPage.mock.results[0].value;
  // Now do the same with `getVersions()`. (We have to do them in sequence
  // because `getVersions()` doesn't get called until `getPage()` finishes.
  await mockApi.getVersions.mock.results[0].value;

  // Now do your expectations, the same as before:
  expect(document.title).toBe('Scanner | http://www.ncei.noaa.gov/news/earth-science-conference-convenes');

  pageDetails.unmount();

  expect(document.title).toBe('Scanner');
});

Reply:
(This is definitely too complicated! Unfortunately, we’ve got all the data loading logic mixed into this component, so to test anything about it, we have to handle the data loading layer. 😕)

Reply:
Thanks so much @Mr0grog!  This was super helpful in better understanding why the tests weren't working how I thought they were and learning a better way to do them.  I didn't even realize GetDerivedStateFromProps was firing off and resetting the page to null because my page id's weren't matching.  Pretty awesome that Jest can mock async calls for the fake data.

Ok, I gave your suggestions a go and they seem to be giving test results as expected 🎉!  To keep consistent with the other test files, I took out your (very helpful) explanatory comments before making my latest commit.

Let me know if there's anything else to take care of with this PR!